packages/flutter_tools/lib/src/_flutter_manifest.dart

```diff
@@ -0,0 +1,460 @@
+    this.assets = const <Uri>[],
+    this.fonts = const <Font>[],
+  });
+
+  /// The string value of the top-level `name` property in the `pubspec.yaml` file.
+  final String appName;
+
+  /// The version String from the `pubspec.yaml` file.
+  /// Can be null if it isn't set or has a wrong format.
+  final String appVersion;
+
+  /// The build version name from the `pubspec.yaml` file.
+  /// Can be null if version isn't set or has a wrong format.
+  final String buildName;
+
+  /// The build version number from the `pubspec.yaml` file.
+  /// Can be null if version isn't set or has a wrong format.
+  final String buildNumber;
+
+  /// True if the project uses Material Design.
+  final bool usesMaterialDesign;
+
+  /// The type of Flutter project.
+  final ProjectType projectType;
+
+  /// The list of assets.
+  final List<Uri> assets;
+
+  /// The list of fonts.
+  final List<Font> fonts;
+
+  @override
+  String toString() {
+    return '$runtimeType(appName: $appName, appVersion: $appVersion, buildName: $buildName, '
+        'buildNumber: $buildNumber, projectType: $projectType)';
+  }
+}
+
+/// A wrapper around each of the sections under `fonts:` in `pubspec.yaml`.
+final class Font {
+  Font(this.familyName, this.fontAssets)
+    : assert(familyName != null),
+      assert(fontAssets != null),
+      assert(fontAssets.isNotEmpty);
+
+  final String familyName;
+  final List<FontAsset> fontAssets;
+
+  Map<String, dynamic> get descriptor {
+    return <String, dynamic>{
+      'family': familyName,
+      'fonts': fontAssets.map<Map<String, dynamic>>((FontAsset a) => a.descriptor).toList(),
+    };
+  }
+
+  @override
+  String toString() => '$runtimeType(family: $familyName, assets: $fontAssets)';
+}
+
+/// A wrapper around each of the nested `fonts:` in `pubspec.yaml`.
+final class FontAsset {
+  FontAsset(this.assetUri, {this.weight, this.style})
+    : assert(assetUri != null);
+
+  final Uri assetUri;
+  final int weight;
+  final String style;
+
+  Map<String, dynamic> get descriptor {
+    final Map<String, dynamic> descriptor = <String, dynamic>{};
+    if (weight != null) {
+      descriptor['weight'] = weight;
+    }
+    if (style != null) {
+      descriptor['style'] = style;
+    }
+    descriptor['asset'] = assetUri.path;
+    return descriptor;
+  }
+
+  @override
+  String toString() => '$runtimeType(asset: ${assetUri.path}, weight; $weight, style: $style)';
+}
+
+/// A service class for creating [FlutterManifest].
+abstract class FlutterManifestService {
+  /// Creates a [FlutterManifest] by parsing the `pubspec.yaml` file in [path].
+  /// Returns [null] if the manifest is invalid.
+  FlutterManifest createFromPath(String path);
+
+  /// Creates a [FlutterManifest] by parsing the [manifest].
+  /// Returns [null] if the manifest is invalid.
+  FlutterManifest createFromString(String manifest);
+}
+
+/// Default implementation of [FlutterManifestService].
+class FlutterManifestServiceImpl extends FlutterManifestService {
```

Comment:
There is no real reason to create Impl classes in Dart, as it doesn't protect you again extending/implementing the interface. The implementation should be moved under the manifest service.

Reply:
For inject-able classes, a library-public abstract class with a library-private Impl class can make sense to avoid internal code relying on library-private members that won't be present in injected implementations. I think we should probably think about adopting this pattern for inject-able classes for this reason.

Reply:
That only protects again code within the same library that is not already part of the implemention using private members. How much of a problem is that really?

We do use this pattern in a few places, but there are other reasons. See codegen.dart for example, we can't import build_runner in google3 so the pattern separates the interface/impl so that the latter can depend on unsupported code.